### DISCUSSION TITLE
Streams

### DISCUSSION THREAD

**Discussion by spunkyvt** [dokan-dev/dokan-dotnet#341] [Q&A] [answered]
I am trying to get the ADS (Alternate Data Streams) working with the Mirror sample in c#.  In the FindStreams I am iterating over the Streams for the file and adding the appropriate values to the out streams variable, but it's only seeing the first one when I use powershell to show the streams.  Also the paths are messed up for the values.  I used the default mirror and it works correctly.  ANy ideas where I need to look next?

**Comment by Olof-Lagerkvist** [comment]
Could you share some code, particularly your `FindStreams` implementation?

**Comment by spunkyvt** [comment]
var fileAttributes = new FileInfo(GetPath(fileName));

            streams = new List<FileInformation>();


            foreach (var item in FileStreamSearcher.GetStreams(fileAttributes))
            {
                var st = new FileInfo(GetPath(fileName + item));

                var newst = new FileInformation();

               newst.Attributes = st.Attributes;
                newst.CreationTime = st.CreationTime;
                newst.LastAccessTime = st.LastAccessTime;
                newst.LastWriteTime = st.LastWriteTime;
                newst.Length = st?.Length ?? 0;
                newst.FileName = item;
                if (item.Contains(":MyXML"))
                    newst.FileName = ":MyXML"; //Added to remove the training :$Data
                streams.Add(newst);
            }


            return Trace(nameof(FindStreams), fileName, info, DokanResult.Success);
            
            
            
            Tracing through it's getting returning a BufferOverflow in the FindStreamsProxy , not sure why

**Comment by spunkyvt** [comment]
It's writing the stream, but if I just query for the Stream in powershell it's missing the MyXML stream

**Comment by Olof-Lagerkvist** [comment]
Are you really sure about what format this `FileStreamSearcher.GetStreams` uses when returning stream names? Do they really begin with a colon like you seem to expect? Have you checked that `MyXML` is really one of the streams enumerated from that call, for example by writing trace output, logging or breakpoints?

**Comment by spunkyvt** [comment]
When using powershell -Stream, it returns 2 streams.  A $DATA and $MyXML.  That is also what is returned when I use the FindFirstStream and FindNextStream API calls ( which is what I use in FileStreamSearcher.GetStreams).  I worked around the issue by only including MyXML instead of trying to write both streams.

**Comment by Olof-Lagerkvist** [comment, marked as answer]
Okay, this is a bit complicated but in short, `$DATA` is not a stream name, it is a stream type. In this case for the main unnamed data stream of the file, the stream that you get if you just open the file without specifying a stream name. If you use `dir /r` at command prompt, you will not see `:$DATA` suffixed to the default streams even though it is technically there. `$DATA` is however suffixed to all other data stream names. In PowerShell it is kind of the other way around, when using `Get-Content -Stream` the output is suffixed with `$DATA` for the main unnamed stream only and not for any other named data stream. I understand that this could be confusing but it is the way it works. So if you use PowerShell as backend for your queries for alternate streams, you need to skip over the stream suffixed with `::$DATA` because that is simply the main, unnamed stream.
